parser-typechecker/src/Unison/DataDeclaration.hs

```diff
@@ -479,6 +496,24 @@ builtinDataDecls = case hashDecls $ Map.fromList
     [ ((), v "Test.Result.Fail", Type.text () `arr` var "Test.Result")
     , ((), v "Test.Result.Ok"  , Type.text () `arr` var "Test.Result")
     ]
+  doc = DataDeclaration
+    (Unique "c63a75b845e4f7d01107d852e4c2485c51a50aaaa94fc61995e71bbee983a2ac3713831264adb47fb6bd1e058d5f004")
+    ()
+    []
+    [ ((), v "Doc.Blob", Type.text () `arr` var "Doc")
+    , ((), v "Doc.Link", Type.ref() linkRef `arr` var "Doc")
+    , ((), v "Doc.Signature", Type.termLink() `arr` var "Doc")
+    , ((), v "Doc.Source", Type.ref() linkRef `arr` var "Doc")
+    , ((), v "Doc.Evaluate", Type.termLink() `arr` var "Doc")
+    , ((), v "Doc.Join", Type.app() (Type.vector()) (var "Doc") `arr` var "Doc")
+    ]
+  link = DataDeclaration
+    (Unique "a5803524366ead2d7f3780871d48771e8142a3b48802f34a96120e230939c46bd5e182fcbe1fa64e9bff9bf741f3c04")
+    ()
+    []
+    [ ((), v "Link.Term", Type.termLink () `arr` var "Link")
+    , ((), v "Link.Type", Type.typeLink () `arr` var "Link")
+    ]
```

Comment:
I had to declare these types in Haskell code because the parser needs them and it would be a circular dependency to put them in `IOSource` (since `IOSource` needs the parser). It would be possible to abstract the parser somewhat to break this cycle but that was more work than I was willing to do.

Reply:
Maybe into a separate Haskell module if DataDeclaration.hs gets out of control.